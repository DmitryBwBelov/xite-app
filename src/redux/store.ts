// Modules
import {applyMiddleware, createStore} from 'redux';
import {persistReducer} from 'redux-persist';
import thunk from 'redux-thunk';
import AsyncStorage from '@react-native-community/async-storage';
import {createEpicMiddleware} from 'redux-observable';
// Redux devtools
import {composeWithDevTools} from 'redux-devtools-extension';
// Root reducer
import rootReducer from '@reducers/index';
// Epics
import {rootEpic} from '@epics/index';
// Initial State
import {initialState} from './state';

const persistConfig = {
    // Root
    key: 'root',
    // Storage Method (React Native)
    storage: AsyncStorage,
    // Whitelist (Save Specific Reducers)
    whitelist: [],
    // Blacklist (Don't Save Specific Reducers)
    blacklist: [],
};

// Middleware: Redux Persist Persisted Reducer
const persistedReducer = persistReducer(persistConfig, rootReducer);
const epicMiddleware = createEpicMiddleware();
const middleware = [epicMiddleware, thunk];

// Redux: Store
const store = createStore(
    persistedReducer,
    // @ts-ignore
    initialState,
    composeWithDevTools(applyMiddleware(...middleware)),
);

epicMiddleware.run(rootEpic);

export {store};
