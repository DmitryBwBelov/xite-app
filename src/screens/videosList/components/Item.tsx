// Modules
import React from 'react';
import {useDeviceOrientation} from '@react-native-community/hooks';
import {useWindowDimensions} from 'react-native';
// Styles
import {
    AuthorWrapText,
    DescriptionWrap,
    ImageWrap,
    MainWrap,
    TitleWrapText,
} from './styles';
// Types
import {ItemType} from './types/Item.type';

export const Item = React.memo<ItemType>(({item}) => {
    const {title, artist, image_url} = item;
    const {width} = useWindowDimensions();
    const deviceOrientation = useDeviceOrientation();

    return (
        <MainWrap $isLandscape={deviceOrientation?.landscape}>
            <ImageWrap
                source={{uri: image_url}}
                $isLandscape={deviceOrientation?.landscape}
                $width={width}
            />
            <DescriptionWrap>
                <TitleWrapText>{title}</TitleWrapText>
                <AuthorWrapText>{artist}</AuthorWrapText>
            </DescriptionWrap>
        </MainWrap>
    );
});
