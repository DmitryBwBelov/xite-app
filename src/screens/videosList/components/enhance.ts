import {connect} from 'react-redux';
import {compose} from 'redux';
// Selectors
import {videosGenresListStructuredSelector} from '@selectors/stucturedSelectors/genresVideosList';
// Actions
import {filterVideosList} from '@actions/videosList';

const mapDispatchToProps = {filterVideosList};
const withConnect = connect(
    videosGenresListStructuredSelector,
    mapDispatchToProps,
);

export const enhance = compose(withConnect);
