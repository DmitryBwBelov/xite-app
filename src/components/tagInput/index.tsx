// Modules
import React, {useCallback, useEffect, useState} from 'react';
import {MainWrap, SearchWrap} from './styles';
import {IndexPath, SelectItem} from '@ui-kitten/components';

interface TagInputInterface {
    setSelectedValues: (values: any[]) => void;
    values: {name: string; id: string | number}[] | [];
}

export const TagInput = React.memo<TagInputInterface>(
    ({values, setSelectedValues}) => {
        const [selected, setSelected] = useState([new IndexPath(-1)]);

        const getTitle = useCallback(() => {
            return selected
                .filter(({row}) => row > -1)
                .map(({row}) => values[row]?.name)
                .join(', ');
        }, [selected, values]);

        useEffect(() => {
            const items = selected.filter(({row}) => row > -1);
            if (items.length >= 1) {
                setSelectedValues(items.map(({row}) => values[row]?.id));
            }
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [selected]);

        return (
            <MainWrap>
                <SearchWrap
                    multiSelect={true}
                    value={getTitle() || 'Select genres'}
                    selectedIndex={selected}
                    onSelect={index => setSelected(index as IndexPath[])}>
                    {values.map(({id, name}) => (
                        <SelectItem key={id} title={name} />
                    ))}
                </SearchWrap>
            </MainWrap>
        );
    },
);
