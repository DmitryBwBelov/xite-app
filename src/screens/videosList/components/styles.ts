import styled from 'styled-components/native';
import {Icon, Button, Select} from '@ui-kitten/components';

export const MainWrap = styled.View<{$isLandscape: boolean}>`
    background-color: #fff;
    border-color: #ccc;
    border-style: solid;
    border-width: 1px;
    margin-right: ${({$isLandscape}) => ($isLandscape ? 10 : 0)}px;
`;

export const ImageWrap = styled.Image<{$isLandscape: boolean; $width: number}>`
    width: ${({$isLandscape, $width}) =>
        ($isLandscape ? $width / 2 : $width) - 20}px;
    height: 150px;
`;

export const DescriptionWrap = styled.View`
    align-items: center;
    margin: 20px 5px;
`;

export const TitleWrapText = styled.Text`
    font-weight: bold;
`;

export const AuthorWrapText = styled.Text``;

export const FilterWrap = styled.View`
    background: #fff;
    padding: 10px;
    margin: 0 0 10px 0;
`;

export const IconWrap = styled(Icon)`
    width: 20px;
    height: 20px;
`;

export const FilterButton = styled(Button)`
    width: 30px;
    align-self: flex-end;
`;

export const SearchWrap = styled(Select)`
    margin: 5px 0;
`;

export const SearchInputWrap = styled.TextInput`
    background: #fff;
    padding: 10px 5px;
`;
