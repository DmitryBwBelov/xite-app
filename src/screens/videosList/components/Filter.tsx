// Modules
import React, {useEffect, useMemo, useState} from 'react';
import {ConnectedComponent} from 'react-redux';
import {IndexPath, SelectItem} from '@ui-kitten/components';
// Types
import {FilterType} from './types/Filter.type';
import {PayloadSearchInterface} from '@interfaces/videosList/PayloadSearch.type';
// Styles
import {FilterWrap, IconWrap, FilterButton, SearchWrap} from './styles';
// Enhancer
import {enhance} from './enhance';
import {TagInput} from '@components/tagInput';

const FilterComponent = React.memo<FilterType>(
    ({items, genres, filterVideosList}) => {
        const [expanded, setExpanded] = useState(false);
        const [searchYear, setSearchYear] = useState(new IndexPath(-1));
        const [selectedYearValue, setSelectedYearValue] = useState('');
        const [selectedTags, setSelectedTags] = useState([]);

        useEffect(() => {
            if (
                searchYear &&
                searchYear?.row >= 0 &&
                searchYear?.row <= items.length
            ) {
                const currentYear = items[searchYear?.row].release_year;
                setSelectedYearValue(`${currentYear}`);
                filterVideosList({year: currentYear} as PayloadSearchInterface);
            }
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [searchYear, items]);

        useEffect(() => {
            if (selectedTags && selectedTags.length >= 1) {
                filterVideosList({
                    tags: selectedTags,
                } as PayloadSearchInterface);
            }
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [selectedTags]);

        const renderFilter = useMemo(() => {
            return (
                <>
                    <SearchWrap
                        value={selectedYearValue || 'Select year'}
                        selectedIndex={searchYear}
                        onSelect={index => setSearchYear(index as IndexPath)}>
                        {items.map(({release_year}) => (
                            <SelectItem
                                key={release_year}
                                title={release_year}
                            />
                        ))}
                    </SearchWrap>
                    <TagInput
                        setSelectedValues={values =>
                            setSelectedTags(values as never[])
                        }
                        values={genres}
                    />
                </>
            );
        }, [items, genres, searchYear, selectedYearValue]);

        return (
            <FilterWrap>
                <FilterButton
                    onPress={() => setExpanded(!expanded)}
                    appearance="ghost"
                    accessoryLeft={() => (
                        <IconWrap name="options-2-outline" fill="#000" />
                    )}
                />
                {expanded && renderFilter}
            </FilterWrap>
        );
    },
);

export const Filter = enhance(
    FilterComponent as ConnectedComponent<React.FunctionComponent, {}>,
);
