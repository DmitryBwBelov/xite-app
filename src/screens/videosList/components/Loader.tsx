// Modules
import React from 'react';
import {useWindowDimensions} from 'react-native';
import {Spinner} from '@ui-kitten/components';
import {ConnectedComponent} from 'react-redux';
// Styles
import {ButtonWrap, ErrorText, LoaderWrap} from '../styles';
// Types
import {LoaderMessageType} from './types/Loader.type';
// Enhancer
import {enhance} from '../enhance';

const LoaderMessageComponent = React.memo<LoaderMessageType>(
    ({fetchVideosList, isPending, isError}) => {
        const {height} = useWindowDimensions();

        return (
            <LoaderWrap $height={height}>
                {isPending && <Spinner size="large" />}
                {isError && <ErrorText>Something went wrong...</ErrorText>}
                {isError && (
                    <ButtonWrap
                        onPress={() => fetchVideosList()}
                        appearance="ghost">
                        Try again
                    </ButtonWrap>
                )}
            </LoaderWrap>
        );
    },
);

export const LoaderMessage = enhance(
    LoaderMessageComponent as ConnectedComponent<React.FunctionComponent, {}>,
);
