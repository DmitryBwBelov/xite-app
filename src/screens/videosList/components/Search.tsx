// Modules
import React, {useEffect, useState} from 'react';
import {ConnectedComponent} from 'react-redux';
import {Input} from '@ui-kitten/components';
// Types
import {FilterType} from './types/Filter.type';
// Enhancer
import {enhance} from './enhance';
import {PayloadSearchInterface} from '@interfaces/videosList/PayloadSearch.type';

const SearchComponent = React.memo<FilterType>(({filterVideosList}) => {
    const [searchQuery, setSearchQuery] = useState('');

    useEffect(() => {
        if (searchQuery && searchQuery?.length >= 1) {
            filterVideosList({query: searchQuery} as PayloadSearchInterface);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [searchQuery]);

    return (
        <Input
            placeholder="Start to type search"
            value={searchQuery}
            onChangeText={nextValue => setSearchQuery(nextValue)}
        />
    );
});

export const Search = enhance(
    SearchComponent as ConnectedComponent<React.FunctionComponent, {}>,
);
