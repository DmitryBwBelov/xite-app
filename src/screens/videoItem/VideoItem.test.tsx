/**
 * @format
 */

import 'react-native';
import React from 'react';
import {Provider} from 'react-redux';
import {ApplicationProvider, IconRegistry} from '@ui-kitten/components';
import {EvaIconsPack} from '@ui-kitten/eva-icons';
import * as eva from '@eva-design/eva';
import {store} from '@redux/store';
import {VideoItem} from './index';
import {render} from '@testing-library/react-native';
// Data
import itemsList from '../../../dataset.json';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

const testItem = itemsList?.videos[0];

const props = {
    navigation: {navigate: jest.fn(), setOptions: jest.fn()},
    route: {params: {item: testItem}},
};

describe('Video Item screen', () => {
    it('Renders correctly', () => {
        const renderedValue = render(
            <Provider store={store}>
                <IconRegistry icons={EvaIconsPack} />
                <ApplicationProvider {...eva} theme={eva.light}>
                    <VideoItem {...props} />
                </ApplicationProvider>
            </Provider>,
        );

        expect(renderedValue).toMatchSnapshot();
    });

    it('Image renders correctly', () => {
        const renderedValue = render(
            <Provider store={store}>
                <IconRegistry icons={EvaIconsPack} />
                <ApplicationProvider {...eva} theme={eva.light}>
                    <VideoItem {...props} />
                </ApplicationProvider>
            </Provider>,
        );

        const image = renderedValue.getByTestId('test_video_item_image');
        expect(image?.props?.source?.uri).toEqual(testItem.image_url);
    });
});
