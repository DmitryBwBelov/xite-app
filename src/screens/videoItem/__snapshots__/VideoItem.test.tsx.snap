// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Item screen Renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": "https://raw.githubusercontent.com/XiteTV/frontend-coding-exercise/679a82b1e7110c16e14412f1debaa118c10078a9/images/501437/images/app/w522_h292.jpg",
        }
      }
      style={
        Array [
          Object {
            "height": 300,
            "width": 750,
          },
        ]
      }
      testID="test_video_item_image"
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginBottom": 10,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 10,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
            },
          ]
        }
      >
        All In
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontSize": 17,
            },
          ]
        }
      >
        Pants Velour
      </Text>
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "#555",
              "color": "#fff",
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 10,
              "paddingBottom": 5,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 5,
            },
          ]
        }
      >
        2014
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
